osquery/sql/sqlite_collation.cpp

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <regex>
+
+#include <osquery/core/flags.h>
+
+#include <sqlite3.h>
+
+namespace osquery {
+/**
+ * @brief Split version string into numeric and alphabetic character segments.
+ */
+static std::vector<std::vector<std::string>> versionSplit(
+    const std::string& version) {
+  std::vector<std::vector<std::string>> result;
+  std::regex ver_pattern("(\\d+|[a-zA-Z]+)");
+  std::sregex_token_iterator iter_end;
+  std::istringstream iss(version);
+  std::string segment;
+
+  // Split version into segments on periods.
+  while (getline(iss, segment, '.')) {
+    std::vector<std::string> seg_pieces;
+    std::sregex_token_iterator iter_begin(
+        segment.begin(), segment.end(), ver_pattern);
+
+    // Split segment into pieces of consecutive alphabetic and numeric
+    // characters.
+    while (iter_begin != iter_end) {
+      seg_pieces.push_back(*iter_begin++);
+    }
+
+    result.push_back(seg_pieces);
+  }
+
+  return result;
+}
+
+/**
+ * @brief Collate version strings. (Only compares alphanumeric characters.)
```

Comment:
> Only compares alphanumeric characters

Is an interesting question -- How can we handle dpkg and rpm versions? dpkg has things like epochs -- `2:1.1.1` is more than `2.2.2.2`, and I assume rpm has similar. 

Given the rules changes, I think it'll be cleaner if we end up registering `VERSION` and `VERSION_DPKG`, but I'm not sure how to handle that in C code.

Reply:
Wrote in support for `version`, `version_arch`, `version_dpkg`, and `version_rhel`.

I added a bunch of test cases from a mix of real package versions, example versioning, and my own written versions. Not sure if I made the linux support the easiest, but I tried to make it simple enough and compatible.